File: Robotics Kinematics and Dynamics Course/Chapter2/Lesson4/Chapter2_Lesson4.java
public class LieGroups {

    public static double[][] hatOmega(double[] w) {
        double wx = w[0], wy = w[1], wz = w[2];
        return new double[][]{
            {0.0,   -wz,   wy},
            {wz,    0.0,  -wx},
            {-wy,   wx,   0.0}
        };
    }

    public static double[] veeOmega(double[][] W) {
        return new double[]{
            W[2][1],
            W[0][2],
            W[1][0]
        };
    }

    public static double[][] so3Exp(double[] omega, double theta) {
        double n = norm(omega);
        if (n < 1e-9) {
            return identity3();
        }
        double[] w = scale(omega, 1.0 / n);
        double[][] W = hatOmega(w);
        double[][] W2 = matMul3(W, W);
        double[][] I = identity3();

        double[][] R = add3(
            add3(I, scale3(W, Math.sin(theta))),
            scale3(W2, 1.0 - Math.cos(theta))
        );
        return R;
    }

    public static So3LogResult so3Log(double[][] R) {
        double trace = R[0][0] + R[1][1] + R[2][2];
        double cosTheta = (trace - 1.0) / 2.0;
        cosTheta = Math.max(-1.0, Math.min(1.0, cosTheta));
        double theta = Math.acos(cosTheta);
        if (theta < 1e-9) {
            return new So3LogResult(new double[]{0.0, 0.0, 0.0}, 0.0);
        }
        double[][] diff = sub3(R, transpose3(R));
        double factor = theta / (2.0 * Math.sin(theta));
        double[][] W = scale3(diff, factor);
        double[] omega = veeOmega(W);
        double n = norm(omega);
        omega = scale(omega, 1.0 / n);
        return new So3LogResult(omega, theta);
    }

    public static Se3LogResult se3Log(double[][] T) {
        double[][] R = new double[3][3];
        double[] p = new double[3];
        for (int i = 0; i < 3; ++i) {
            for (int j = 0; j < 3; ++j) {
                R[i][j] = T[i][j];
            }
            p[i] = T[i][3];
        }
        So3LogResult logR = so3Log(R);
        double[] omega = logR.omega;
        double theta = logR.theta;

        double[] v;
        if (theta < 1e-9) {
            theta = norm(p);
            if (theta < 1e-9) {
                v = new double[]{0.0, 0.0, 0.0};
                omega = new double[]{0.0, 0.0, 0.0};
                return new Se3LogResult(v, omega, 0.0);
            }
            v = scale(p, 1.0 / theta);
            omega = new double[]{0.0, 0.0, 0.0};
            return new Se3LogResult(v, omega, theta);
        }

        double[][] W = hatOmega(omega);
        double[][] W2 = matMul3(W, W);
        double[][] I = identity3();

        double c = Math.cos(theta);
        double s = Math.sin(theta);
        double[][] Jinv = add3(
            add3(I, scale3(W, -0.5)),
            scale3(
                W2,
                (1.0 / (theta * theta)) - (1.0 + c) / (2.0 * theta * s)
            )
        );

        v = matVec3(Jinv, p);
        return new Se3LogResult(v, omega, theta);
    }

    // Helper classes and vector/matrix utility functions omitted for brevity:
    // norm, scale, identity3, matMul3, matVec3, add3, sub3, transpose3, scale3, etc.
    // They should implement standard 3x3 operations.
}
      

File: Robotics_Kinematics_and_Dynamics_Course/Chapter2/Lesson4/Chapter2_Lesson4.m
function demo_se3_exp_log()
    omega = [0; 0; 1];
    theta = pi/4;
    v = [0.1; 0.0; 0.0];
    xi = [v; omega];

    T = se3_exp(xi, theta);
    [xi_rec, theta_rec] = se3_log(T);

    disp('T ='); disp(T);
    disp('Recovered xi and theta:');
    disp(xi_rec.'); disp(theta_rec);
end

function W = hat_omega(omega)
    wx = omega(1); wy = omega(2); wz = omega(3);
    W = [  0,   -wz,   wy;
          wz,    0,   -wx;
         -wy,   wx,    0 ];
end

function [omega, theta] = so3_log(R)
    cos_theta = (trace(R) - 1) / 2;
    cos_theta = max(-1, min(1, cos_theta));
    theta = acos(cos_theta);
    if theta < 1e-9
        omega = [0; 0; 0];
        theta = 0;
        return;
    end
    W = (theta / (2 * sin(theta))) * (R - R');
    omega = [W(3, 2); W(1, 3); W(2, 1)];
    omega = omega / norm(omega);
end

function R = so3_exp(omega, theta)
    n = norm(omega);
    if n < 1e-9
        R = eye(3);
        return;
    end
    omega = omega / n;
    W = hat_omega(omega);
    W2 = W * W;
    R = eye(3) + sin(theta) * W + (1 - cos(theta)) * W2;
end

function T = se3_exp(xi, theta)
    v = xi(1:3);
    omega = xi(4:6);
    T = eye(4);
    if norm(omega) < 1e-9
        % pure translation
        T(1:3, 4) = v * theta;
        return;
    end
    omega = omega / norm(omega);
    W = hat_omega(omega);
    W2 = W * W;
    R = so3_exp(omega, theta);
    I3 = eye(3);
    J = I3 * theta + (1 - cos(theta)) * W + (theta - sin(theta)) * W2;
    p = J * v;
    T(1:3, 1:3) = R;
    T(1:3, 4) = p;
end

function [xi, theta] = se3_log(T)
    R = T(1:3, 1:3);
    p = T(1:3, 4);

    [omega, theta] = so3_log(R);
    if theta < 1e-9
        theta = norm(p);
        if theta < 1e-9
            xi = zeros(6, 1);
            theta = 0;
            return;
        end
        v = p / theta;
        xi = [v; 0; 0; 0];
        return;
    end
    W = hat_omega(omega);
    W2 = W * W;
    I3 = eye(3);
    Jinv = I3 ...
           - 0.5 * W ...
           + (1 / (theta^2) - (1 + cos(theta)) / (2 * theta * sin(theta))) * W2;
    v = Jinv * p;
    xi = [v; omega];
end
      
